### ISSUE TITLE
theming: New CSS-Variables for Shellbar & Toolbar Height

### ISSUE THREAD

**Issue by C3ntraX** [SAP/ui5-webcomponents#7197]
### Feature Request Description

I would appreciate to get access for two new css-variables, as they are not offically supported but I already had references to it:

The variables I was using in previous versions:
`--_ui5_shellbar_root_height` - To determine the shellbar height
`--sapWcrToolbarHeight` - To determine the toolbar height

Rename this variables if you needed.

Both variables are private. `--sapWcrToolbarHeight` is coming from react, I don't know why.
The naming in react --sap* is currently private. There is a issue open to add the '_' to make it clear.

**Why I need this variables**

The main reason why they should be accessible via css-variables and not via other ways like the workarounds:
`getComputedStyle(<ui5-shellbar>).getPropertyValue('--_ui5_shellbar_root_height')`
or via getBoundingClientRect
`document.getElementById('shellbar').getBoundingClientRect(). height`

Is the **performance** and easy **accessibility**

See: #7182

TLDR:
For shellbar: https://codesandbox.io/p/sandbox/css-variables-n4r7m5?file=%2Fsrc%2FApp.tsx%3A10%2C5-19%2C21

>   <ThemeProvider>
      <ShellBar
        data-testid="shell-wrapper"
        style={{ position: "fixed", width: "100%", zIndex: 100 }}
      />
      <div
        data-testid="emptySpace-wrapper"
        style={{ paddingTop: "var(--_ui5_shellbar_root_height)" }}
      />
    </ThemeProvider>

For Toolbar, I use the height to calculate the top empty space for toasts.
The Shellbar height is also used to calculate the top emtpy space for toasts.

### Proposed Solution

Renamed variables of:
`--_ui5_shellbar_root_height`
`--sapWcrToolbarHeight`

https://github.com/SAP/theming-base-content

### Priority

Medium

### Privacy Policy

- [X] I’m not disclosing any internal or sensitive information.

**Comment by nnaydenow** [comment]
Hi @C3ntraX,

Could you check following snippix (https://jsbin.com/lusewulaqu/edit?html,output)? It simulates the layout that you want to achieve. The red area is for shellbar position (height for element with class .child is not required since it will automatically take the height from component) and the light blue area is for the content and it will take the remaining space.

With this solution I think that usage of following variables is not necessary.

**Comment by C3ntraX** [comment]
Hi @nnaydenow,

the important part is the accessibility for the whole react app.

Additinal, your senario does not work for me. Because of position: "fixed".

If you want to see how many codes of line does I needed for this to work (which is not performant):

useRect.tsx
```
// source: https://gist.github.com/morajabi/523d7a642d8c0a2f71fcfa0d8b3d2846
import { useLayoutEffect, useCallback, useState } from "react";

type RectResult = {
   bottom: number;
   height: number;
   left: number;
   right: number;
   top: number;
   width: number;
};

function getRect<T extends HTMLElement>(element?: T): RectResult {
   let rect: RectResult = {
      bottom: 0,
      height: 0,
      left: 0,
      right: 0,
      top: 0,
      width: 0,
   };
   if (element) rect = element.getBoundingClientRect();
   return rect;
}

export function useRect<T extends HTMLElement>(ref: React.RefObject<T>): RectResult {
   const [rect, setRect] = useState<RectResult>(ref && ref.current ? getRect(ref.current) : getRect());

   const handleResize = useCallback(() => {
      if (!ref.current) return;
      setRect(getRect(ref.current)); // Update client rect
   }, [ref]);

   useLayoutEffect(() => {
      const element = ref.current;
      if (!element) return;

      handleResize();

      // @ts-ignore
      if (typeof ResizeObserver === "function") {
         // @ts-ignore
         let resizeObserver: ResizeObserver | null = new ResizeObserver(() => handleResize());
         resizeObserver.observe(element);
         return () => {
            if (!resizeObserver) return;
            resizeObserver.disconnect();
            resizeObserver = null;
         };
      } else {
         window.addEventListener("resize", handleResize); // Browser support, remove freely
         return () => window.removeEventListener("resize", handleResize);
      }
   }, [ref.current]);

   return rect;
}
```

Shell.tsx
```
import { useRect } from "../../hooks/useRect";
import { setShellbarRootHeight } from "../../app/slices/uiSlice";
```

```
import {
   ShellBarDomRef,
} from "@ui5/webcomponents-react";
...
   const shellbarRef = useRef<ShellBarDomRef>(null);
   const { height } = useRect(shellbarRef);
   useEffect(() => {
      if (height) {
         dispatch(setShellbarRootHeight(height));
      }
   }, [dispatch, height]);

return <UI5Shellbar
  ...
  ref={shellbarRef}
  ... />
<div data-testid="emptySpace-wrapper" style={{ paddingTop: height }} />
```

Redux: uiSlice.ts
```
setShellbarRootHeight: (state, { payload: height }: PayloadAction<number>) => {
         state.shellbarRootHeight = height;
 },

```

CSSBaseLine.tsx
Global css variable (should't need this)
```
const CSSBaseLine = () => {
   const shellbarRootHeight = useTypedSelector((state) => state.ui.shellbarRootHeight);

   useEffect(() => {
      document.documentElement.style.setProperty("--pageHeight", "100%");
      if (shellbarRootHeight) {
         document.documentElement.style.setProperty("--shellbar_root_height", `${shellbarRootHeight}px`);
      }
   }, [shellbarRootHeight]);

   return null;
};
```

```
ToastWrapper.styles.ts
root: {
      top: (theme: ToastWrapperTheme) => {
         // const toolbars =
         //    theme.pathname.includes("dashboard") || theme.pathname.includes("login") || theme.pathname === "/" ? 0 : 2;
         const toolbars = 0;
         const extraTop = toolbars > 0 ? "10px" : "0px";

         return `calc(${toolbarHeight} + (var(--sapWcrToolbarHeight) * ${toolbars}) + ${extraTop} + ${marginTop})`;
      },
   },
```


This is `very very` much code for just one/two variable(s). Please, I hope you understand now, that this is not to underestimate. You always have to listen for height changes or other listeners to setup. And to make it global availale, you want a global place to setup your variables. I mean I could create my css variable in the redux slice but this is a better approch.

**Comment by ilhan007** [comment]
@C3ntraX what Nayden suggested is not with position: fixed, it's combination of using display: flex, flex-grow and overflow: scroll, to have the ShellBar on top and scrolling container below, without the need of knowing the height of the ShellBar.
It's a pure CSS solution, none of the code of getting the height that you showed is required.

Apart of the example/the solution Nayden gave, we discussed the option to make some of the internal CSS Vars public. Unfortunately we decided to keep them private, because we would like to have the freedom the change them, otherwise it would be breaking change.
The only CSS Vars that yo can use and consider public are these:  https://github.com/SAP/theming-base-content/blob/master/content/Base/baseLib/sap_horizon/css_variables.css

**Comment by C3ntraX** [comment]
Bad decision

**Comment by C3ntraX** [comment]
Some css vars like I want will never be a braking change. Beaucae they are always needed. So I really don't understand your solution to make the developers experience bad!

**Comment by ilhan007** [comment]
Hi @C3ntraX 

1. **renaming the variable and/or removal of the variable will be a breaking change**. If it was not breaking, you would not file an issue, we moved the variable definition from the root to the custom element itself and that broke you app, isn't it?

2. We also suggested a clean approach with pure CSS to create the layout - `without position: fixed`, `without the need of knowing the height` of our Shellbar, with the use of `display: flex`, and  `flex-grow` 
https://codesandbox.io/s/ui5-webcomponents-forked-l94pgc

3. The reason we moved the variables from the root to be scoped on the element is because multiple versions of UI5 Web Components could be used in a same app, and then on root level, CSS Variables are clashing (vars defined from older runtime with vars defined in newer), leading to visual issues. 
If you are interested, you can get familiar with this change:  https://github.com/SAP/ui5-webcomponents/pull/7044 and issue that was addressed https://github.com/SAP/ui5-webcomponents/issues/6876

4. We don't make a bad development experience, we are just explaining that what you do is not the best way,
 you don't need to use our private variables and must not rely on our private variables, there are better ways to create your layout with pure CSS, without any JS involved.

BR,
ilhan
